### ISSUE TITLE
Singularize `as` in foreign keys in To-Many associations (amendment of PR #5957)

### ISSUE THREAD

**Issue by overlookmotel** [sequelize/sequelize#6023]
PR #5957 recently changed the names of foreign keys in associations where an `as` alias is specified.

So now: `Project.hasMany(User, {as: 'Workers'})` creates foreign key `WorkersId` rather than `UserId`.

This is great, but I do think the `as` should be singularized, so foreign key is `WorkerId`. This was already raised by @mickhansen in #5957, but I was too late to the party to get a comment in.

`WorkersId` (plural), in my view, is not what the user would expect. It's inconsistent with the naming when no `as` alias is provided, and with the general Sequelize convention of singularizing/pluralizing. I can imagine it tripping people up quite often because it's counter-intuitive.

Yes, using `inflection` to singularize `as` could in some few cases produce the wrong result, but in those cases the user can override it by specifying `foreignKey`. Personally, I think that's preferable to having to specify `foreignKey` in all cases in order to maintain consistent naming.

I'd be happy to make a PR for this, if there's agreement that this would be a good thing to do.

What do you think @mickhansen @janmeier @sushantdhiman?

**Comment by sushantdhiman** [comment]
We should avoid magically transforming the name to something else, It may cause confusion once again. If user want a specific `foreignKey` they can always explicitly define it. But our transformation should not give any surprises.

**Comment by mickhansen** [comment]
@overlookmotel I definitely agree that `WorkersId` is extremely ugly - But i also remember what a huge pain in the ass inflection has been in the past, it makes it much harder for us to garantee SEMVER when inflection can break in so many subtle ways.

The problem is that it basically forces any reasonable person to ALWAYS use `foreignKey` when using `as` (because most people deploying to production should care how there database looks).

**Comment by overlookmotel** [comment]
Yes I remember being caught out with inflection in the past! But inflection is still being used for pluralization within Sequelize e.g. with one-to-many relationships `User.findAll({include: Order})` returns the orders as `Orders` (plural). So Sequelize is dependent on inflection already.

I totally agree that obeying semver is important. I'd suggest the way to guarantee that is to pin the version of inflection and only update at major version bumps of Sequelize.

@sushantdhiman:
> But our transformation should not give any surprises.

I agree. But I think getting a foreign key `WorkersId` *is* surprising!

I totally see that there are arguments on both sides and of course I go with the consensus. But I do feel that this could be the cause of users posting issues "why am I getting `WorkersId`?" which is going to be an annoyance.

In your experience, how often does inflection get it wrong? If the answer is like 1% or 5%, I'd argue it's best to cater for the majority of cases and keep it consistent with the rest of Sequelize, rather than have a behaviour which is 100% predictable but always weird!

**Comment by mickhansen** [comment]
I'm leaning to your side @overlookmotel since i find it kind of horrible that `foreignKey` will be defacto required if you use `as`.

/cc @janmeier

**Comment by janmeier** [comment]
Thinking this over again I guess singularizing `as` by default is going to give us the least amount of headackes

**Comment by mickhansen** [comment]
4.0.0 is still a few weeks off (atleast), so there's time for more thought and discussion on the matter :)

**Comment by overlookmotel** [comment]
@sushantdhiman what do you think?

If everyone is in agreement for singularization, it'd be good if I got on the case with a PR sooner rather than later. Personally I'd like to start building things with Sequelize v4 soon to give it a bit of a road test before release, and so it'd be preferable if breaking changes were decided on (even if not implemented yet!).

But if anyone wants to think it over further, that's another matter...

**Comment by overlookmotel** [comment]
Does anyone have a guess to how often inflection gets it wrong? Like 1% or 20%? In my experience probably more like 1-2%.

**Comment by sushantdhiman** [comment]
I would still prefer to keep the foreign key to match `as` to avoid any auto magical transformations. Begin certain 100% time is better than 95% time.

Writing an additional `foreignKey` should not be that hard. How many associations using `as` are defined in a normal project may be 10 or 20, mostly in same `index.js` file. Defining a `foreignKey` (when required) should not be difficult. I think it will bring more clarity to me if I can see all the `foreignKey` directly in `index.js`

@overlookmotel : 
> I agree. But I think getting a foreign key WorkersId is surprising!

I think being inflected on my defined name is surprising to me. Different users will have different perspectives.

Its more clear if we say, whatever you put in `as` will be used to form the `foreignKey` rather than we will try to inflect it, but we don't know what name we are going to use.

@mickhansen @janmeier @overlookmotel 

I will have no problems with whatever decision you guys take,I am sure It will be better for everyone. But if you want to use inflections just because of ugly names (which can be avoided by little developer efforts) please reconsider.

**Comment by mickhansen** [comment]
> Writing an additional foreignKey should not be that hard

It's not, but we might aswell throw an error for undefined foreignKey, since `as` on it's own would be very poor practive.

> I think being inflected on my defined name is surprising to me. Different users will have different perspectives.

We do know what we inflect, it would be the alias.

**Comment by sushantdhiman** [comment]
> It's not, but we might aswell throw an error for undefined foreignKey, since as on it's own would be very poor practive.

Thats a good option, but should we force this upon users? If we keep using `as` for naming, there is no need for that. If user still want to change the `foriegnKey` they are free to do so.

> We do know what we inflect, it would be the alias.

But can we guarantee it will be correctly converted

**Comment by felixfbecker** [comment]
An `id` always points to **one** row, so it should be `WorkerId`, not `WorkersId`. Inflection is used in other places too, so this should be the default, and @mickhansen is right that any sane person uses explicit names anyway. I don't find it surprising that the alias is inflected for the column name as I only see the alias as the Model key that holds the array when including. In my (relational) database I would expect a singular, because the id points to **one** row.

**Comment by mickhansen** [comment]
> Thats a good option, but should we force this upon users? If we keep using as for naming, there is no need for that. If user still want to change the foriegnKey they are free to do so.

Non singularized foreign key names for plural associations would be a bad standard to set, and i think we should attempt to guide users to the correct path as much as possible.

> But can we guarantee it will be correctly converted

Correct, but i'm not sure this is better than the alternative :)

Most serious developers will set `foreignKey` for any long lived project. But for newer developers or hobby projects i think we should still attempt to enforce good practices (for 5.0 i'll probably look at introduce naming strategies and change the default one).

**Comment by overlookmotel** [comment]
OK, just to rewind a bit, here's my summary of discussion so far:

1. I think everyone agrees that singularizing `as` would be ideal if there was a foolproof way to do it.
2. The concern is that [inflection](https://www.npmjs.com/package/inflection) does not singularize with 100% accuracy.
3. Consistency needs to be maintained so that behavior doesn't change within a major release of Sequelize (i.e v4.9.1 behaves same as v4.0.0).

Now my personal view:

I think that [inflection](https://www.npmjs.com/package/inflection) is accurate *enough*. It's come a long way in past couple of years. It has 600,000 downloads a month, [no open issues](https://github.com/dreamerslab/node.inflection/issues), and the maintainer is very quick to resolve issues that do arise. They now use [Wiktionary](https://en.wiktionary.org/) as a reference, so questions of whether behavior is correct or not can be easily resolved. I think it's safe to assume that its now working correctly for all but the most obscure words.

Personally (and you may think I'm mad) I don't want to define `foreignKey` for every association. This is something I want (and expect) my ORM to do for me.

To deal with the need for consistency I propose 3 things:

1. Update `inflection` to latest before release of Sequelize v4 and pin it so it does not change until v5.
2. Ensure that Sequelize codebase only references `Utils.inflection` rather than referring to `inflection` directly. Then, if a user finds a bug in inflection that affects their app and they want to update it to a later version, they can use `Sequelize.Utils.inflection = require('inflection')` to inject a later version.
3. The above restriction can be enforced in the test shims to throw an error if inflection is ever accessed directly.

The above would I think be the best route to do the "right" thing in 99% of cases and offer an easy, consistent way for users to deal with the 1%.

**Comment by janmeier** [comment]
+1 on the through research on inflection @overlookmotel - Given your arguments I definitely think we should trust inflection to do the right thing.

Locking the version might be a good idea - On one hand you can argue that inflection is always right, but even an update that makes something right might break user code if the depend on the wrong behaviour. I think I agree that allowing users to inject a newer version if the need it would be a good way to go

**Comment by felixfbecker** [comment]
@janmeier @overlookmotel does inflection release changed inflection behaviour as a major, minor or patch?

**Comment by felixfbecker** [comment]
https://github.com/dreamerslab/node.inflection/blob/master/History.md

**Comment by overlookmotel** [comment]
@felixfbecker All corrections are regarded as bug fixes so may happen on a minor release (though the maintainer *mostly* bumps minor version).

As @janmeier said, regardless of whether it's a "bug fix" or not, it's a change of behavior. I want to be able to update Sequelize within v4.x.x and know for sure (or as sure as possible) that nothing's going to break. So in this case I think consistency is more important than having the latest patch.

**Comment by mickhansen** [comment]
Lock at batch + write an interface around inflection so it's changeable by the end user/plugin seems like a good idea.

In general i'd like a user or plugin to be able to inject a general naming strategy used across table names, foreign key names, attribute fields, etc.

**Comment by felixfbecker** [comment]
@mickhansen Yeah. Like for example I find it very weird that foreign keys are uppercased, like `UserId` (in the table and in the model!). Properties should always be camelCase. Only classes should be PascalCase.

**Comment by mickhansen** [comment]
@felixfbecker I personally agree, however the naming convention is before either me or Jan joined as maintainers :)

My personal style is camelCase attributes / model names, snake_case table names / field names

**Comment by felixfbecker** [comment]
@mickhansen exactly my style too and that also seems to be the convention. Currently we set `foreignKey` and `as` everywhere in our app to enforce the lowercase attribute name, it would be nice to have an option for this (just like you can set `underscored`). The snake_case is easier to read and DBMS are not always case sensitive, for example MySQL ist lowercase by default on Windows and PostgreSQL will lowercase your identifiers if you don't quote them. It is better to simply always use snake_case.

**Comment by overlookmotel** [comment]
OK, how about I submit a PR for the specific change I proposed (singularize foreign keys) and then we work out an API for more broadly allowing the user to specify a naming convention to be implemented later?

**Comment by mickhansen** [comment]
Fine by me

**Comment by overlookmotel** [comment]
I've run into some problems with this. The current behavior doesn't make sense to me, and I think the whole idea of using `as` to create foreign keys has gone a bit wonky.

A couple of examples from the tests in latest master:

#### hasMany

```js
var City = sequelize.define('City', { name: DataTypes.STRING }),
    State = sequelize.define('State', { name: DataTypes.STRING });

State.hasMany(City);
// created field `StateId` on `City` table - good!

State.hasMany(City, {as : 'Capital'});
// created field `CapitalId` on `City` table - huh?
```

This doesn't make sense to me - the City *is* the Capital. So it makes no sense for the ID field referring to the `State` to be called `CapitalId`. The old behavior, where the field was called `StateId` made more sense.

#### belongsToMany

```js
var User = sequelize.define('user', {});
var UserFollowers = sequelize.define('userFollowers', {
    id: { type: DataTypes.INTEGER, primaryKey: true, autoIncrement: true }
}, { timestamps: false });

User.belongsToMany(User, {
    as: 'Followers',
    through: UserFollowers
});
// created fields `FollowerId` and `FollowersId` on `UserFollowers` table
```

This is just confusing more than anything. With `FollowerId` and `FollowersId`, it's not clear which field represents which direction of the association. And of course, if you singularize (as I was planning to do), you'd end up with both fields called `FollowerId` which obviously doesn't work.

The old behavior for many-to-many self-associations was to have fields `UserId` and `FollowerId`.

I'm now thoroughly confused. In making the foreign key from `as`, what was the desired behavior?

**Comment by overlookmotel** [comment]
Also, a more general idea about associations here: https://github.com/sequelize/sequelize/issues/6091

**Comment by felixfbecker** [comment]
@overlookmotel I think the examples you give are wrong. Shouldn't a state have only one capital? In that case, it would be
```js
State.hasMany(City) // creates stateId on City
State.hasOne(City, {as: 'capital'}) // creates capitalId on State
```
Your example with `belongsToMany` is a special case because you are associating a model with itself. Also, you don't need to define an extra model for the relation, why should the UserFollowers have an id? The primary key is a compound key over both foreign keys. Just do:
```js
User.belongsToMany(User, {as: 'followers', through: 'UserFollowers'})
```
I haven't tried that out but I would expect it to add a table `UserFollowers` with `userId` and `followerId`.

**Comment by mickhansen** [comment]
@felixfbecker I personally never use compound primary keys, i always have an `id` on my join tables (my join tables often have extra data aswell though)

**Comment by mickhansen** [comment]
I'd actually tend to agree with @overlookmotel here, although i reviewed and accepted the other PR.

`Parent.hasMany(Children, {as: 'Children'})` will result in child having a childrenId column rather than parentId.

I knew there was a reason we only had it for `belongsTo`

**Comment by felixfbecker** [comment]
@mickhansen for additional data you have to define another model of course and pass it to through. But how do you enforce that you cannot associate the same two rows twice? Do you use a compound unique constraint? I find a compound primary key much easier.

`Parent.hasMany(Child, {as: 'children'})` should of course create `parentId` on `Child`.

**Comment by mickhansen** [comment]
@felixfbecker Yes, i never use `through` as a string. I use a PK + a unique constraint normally yes.
@felixfbecker But it doen't if `hasMany` always uses `as`, which we changed recently - But i have a feeling we probably shouldn't have - But that's why we're doing prereleases :)

**Comment by felixfbecker** [comment]
@mickhansen We have about 40 models and one file per model, that would be a lot of extra files if we defined an extra model for every m:n relation :D

I think we all agree that the new behaviour is not what one would expect.

**Comment by overlookmotel** [comment]
@felixfbecker said:

> I think the examples you give are wrong

Yes I totally agree with you. But these examples are taken directly from the tests. I didn't make them up!

But if you put aside the (quite correct) point that state-capital relationship would be one-one rather than one-many (let's assume that in this strange country, every state has 2 capitals), then I think my point stands - the `foreignKey` being `CapitalId` makes no sense.

**Comment by felixfbecker** [comment]
@overlookmotel I agree. The `as` alias is for the _source_ model and should not be used when infering the foreign key on the _target_ model.

**Comment by overlookmotel** [comment]
@felixfbecker Precisely! Why are you able to say things in 10 words when it takes me 200?

**Comment by overlookmotel** [comment]
My god this has opened a whole can of worms!

I've looked into this further and here's what I've found. (sorry that this is really long)

### One-to-many

The foreign key, if it's going to be based on `as`, needs to be based on the *other* `as`. i.e. the `as` defined in the matching `belongsTo` statement.

```js
State.hasMany( City, { as : 'Capital' } );
City.belongsTo( State, { as: 'CapitalOf' } );
```

So instead of `StateId` the foreign key becomes `CapitalOfId`. This makes more sense!

That's actually what `City.belongsTo( State, { as: 'CapitalOf' } )` already does in Sequelize v3.

However, if you define both `hasMany` and `belongsTo` as in the example above, Sequelize v3 doesn't match the two statements as being two sides of the same association and you end up with two fields on `City`: `StateId` and `CapitalOfId`.

The only way to link them together is to define the same `foreignKey` on both.

```js
State.hasMany( City, { as: 'Capital', foreignKey: 'capitalOfId' } );
City.belongsTo( State, { as: 'CapitalOf', foreignKey: 'capitalOfId' } );
```

Personally I don't like this - I feel that the purpose of an ORM is to abstract away the database (including the column names) so I don't feel that it's good to be forced to used `foreignKey` to link `hasMany` and `belongsTo` statements.

But there has to be some element in common between the 2 statements, otherwise Sequelize can't possibly know which statements to link together with which. This is one reason why I think defining a two-way association with a single statement would be simpler (see #6091).

But anyway... I think:

* Previous behavior concerning `as` for one-to-many was better than what we have on latest master
* Best thing is to roll back @sushantdhiman's change for `hasMany`

### Many-to-many

#### Sequelize v3 behavior

```js
// people are members of many groups, groups have many members
Person.belongsToMany( Group, { through: 'PersonGroup' } );
Group.belongsToMany( Person, { through: 'PersonGroup' } );
// creates model `PersonGroup` with fields `PersonId` and `GroupId`

// with `as`
Person.belongsToMany( Group, { as: 'MyGroups', through: 'PersonGroup' } );
Group.belongsToMany( Person, { as: 'Members', through: 'PersonGroup' } );
// creates model `PersonGroup` with fields `PersonId` and `GroupId`
```

i.e. the `as` is not used at all for creating the keys. But it all works.

If you define `foreignKey`, the behaviour is (to my eyes) a bit wonky:

```js
Person.belongsToMany( Group, { through: 'PersonGroup', foreignKey: 'GroupId' } );
Group.belongsToMany( Person, { through: 'PersonGroup', foreignKey: 'MemberId' } );

var bob = yield Person.create( { id: 1, name: 'Bob' } );
var fighters = yield Group.create( { id: 1, name: 'Fighters' } );
var lovers = yield Group.create( { id: 2, name: 'Lovers' } );

yield bob.addGroup( lovers ); // Bob is a lover not a fighter
// creates row in PersonGroup with Bob's ID as `GroupId` - should be `MemberId`
```

To make it work as expected, you have to switch the two `foreignKey` declarations:

```js
Person.belongsToMany( Group, { through: 'PersonGroup', foreignKey: 'MemberId' } );
Group.belongsToMany( Person, { through: 'PersonGroup', foreignKey: 'GroupId' } );
```

This seems to me counter-intuitive.

#### Current master

Current master uses `as` and creates `MembersId` and `MyGroupsId`. This has two problems:

1. Pluralized keys
2. The two keys are the wrong way round - `MyGroupsId` replaces `PersonId` not `GroupId`. So if you create a relationship between a Person "Bob" and a Group, you end up with Bob's ID saved as `MyGroupsId` where it should be in `MembersId`

#### Solution

We could either:

1. revert to Sequelize v3 behavior, or
2. ignore `as` but switch the `foreignKey`s round, or
3. use the `as` for *both* field names, so the fields become `MemberId` and `MyGroupId` (but the right way around)

Concerning whether to use `as`, I'm not bothered either way. For one-to-one or one-to-many relationships, using `as` as the base for the foreign key is necessary because you can have multiple relationships between the same two entities.

```js
Person.belongsTo( Group, { as: 'Chief' } );
Person.belongsTo( Group, { as: 'Deputy' } );
// creates keys `ChiefId` and `DeputyId` - obviously two fields called `PersonId` wouldn't work!
```

But with many-to-many relationships, each association creates a new through table, so there's no possibility of two foreign keys being called the same thing in the same table.

So I kind of see using `as` as a solution to a problem that doesn't really exist.

However:

* Either way, the behavior in current master is clearly wrong
* In my opinion Sequelize v3 behavior is weird too

### Many-to-many self-association

In a many-to-many *self-*association, providing `as` and `through` is mandatory.

```js
Person.belongsToMany( Person, { as: 'Parents', through: 'PersonParent' } );
Person.belongsToMany( Person, { as: 'Children', through: 'PersonParent' } );
```

#### Sequelize v3

In Sequelize v3, the behavior is broken:

* The first `belongsToMany` creates `PersonParent` model with fields `PersonId` and `ParentId`
* The 2nd `belongsToMany` deletes the `ParentId` field! So you only end up with a single `PersonId` field in the join table.

To get it to work, you have to also specify `foreignKey` in both directions. But, again, the foreign keys have to be what seems to me to be the *wrong way around*:

This works:

```js
Person.belongsToMany( Person, { as: 'Parents', through: 'PersonParent', foreignKey: 'ChildId' } );
Person.belongsToMany( Person, { as: 'Children', through: 'PersonParent', foreignKey: 'ParentId' } );

var bob = yield Person.create( { id: 1, name: 'Bob' } );
var frank = yield Person.create( { id: 2, name: 'Frank' } );
yield bob.addParent(frank); // Frank is Bob's dad
// creates row in PersonParent with Bob's ID as `ChildId` and Frank's ID as `ParentId`
```

#### Current master

```js
Person.belongsToMany( Person, { as: 'Parents', through: 'PersonParent' } );
Person.belongsToMany( Person, { as: 'Children', through: 'PersonParent' } );
```

Foreign keys now don't need to be provided. `ParentsId` and `ChildrenId` fields are created from the `as` - good stuff. BUT the fields are the wrong way around.

```js
var bob = yield Person.create( { id: 1, name: 'Bob' } );
var frank = yield Person.create( { id: 2, name: 'Frank' } );
yield bob.addParent(frank); // Frank is Bob's dad
// creates row in PersonParent with Bob's ID as `ParentsId` and Frank's ID as `ChildrenId` - should be reversed
```

So I think for many-to-many self-associations, both Sequelize v3 and current master have it wrong.

### Conclusions

All of the above is why I say that defining associations in Sequelize is confusing!

But one thing is for sure - the behavior on current master is wrong.

I see 3 possible ways forward:

1. Revert @sushantdhiman's PR and return to Sequelize v3 behavior. Personally I think it's still wacky for many-to-many associations, but at least this doesn't break BC.
2. Try to fix Sequelize v3 behavior for many-to-many (and possibly use `as` as base for foreign key names). This would be a breaking change and I think will be confusing to old users who've got used to it, and will find changing over a head-scratcher.
3. Totally overhaul associations, as suggested in #6091. A big breaking change, but at least there'd be new syntax so it'd be easier to get your head around this being "a new way", rather than old code that looks right but suddenly doesn't do what it used to.

Thoughts?

**Comment by mickhansen** [comment]
I don't think i agree with switching the N:M foreign keys around, it would make it break the logic of hasOne/hasMany saying that the foreign key holds the value pointing back at the source.

**Comment by overlookmotel** [comment]
@mickhansen OK, I see the logic. Maybe it's just a case of making the docs clearer. I'm a *reasonably* smart guy, but have always had trouble getting my head round it. I think if the docs made the logic that the foreign key refers to the target, that'd clear it up.

That said, if I'm the only one who finds it confusing, perhaps the only conclusion is that there's nothing unclear about it, and I'm an idiot!

A few questions remain:

1. Should `as` be used as a base for creating `foreignKey` in many-to-many joins? Or should it be left as was in Sequelize v3? (NB using `as` might add confusion as `as` refers to the source, and `foreignKey` the target)
2. How to fix the broken behavior of many-to-many self-associations? i.e. that the 2nd `belongsToMany()` statement deletes one of the keys from the through table, unless `foreignKey` is provided.

**Comment by sushantdhiman** [comment]
1. `as` behavior should be in sync, otherwise we will have to reopen https://github.com/sequelize/sequelize/issues/5267 and/or find better solution

**Comment by mickhansen** [comment]
@sushantdhiman I'm afraid it doesn't really make sense to have them in sync, foreignkey based on alias only works when the foreign key is describing the target

**Comment by overlookmotel** [comment]
@sushantdhiman As far as I can see, using `as` as basis of `foreignKey` in `hasOne` and `hasMany` makes no sense. Because the foreign key corresponds to the `as` of the matching `belongsTo`. Please see my (loooong) post above where I'm talking about `hasMany` and `as: 'Capital'`.

I think it's an open question whether `as` should be used for `belongsToMany`. I don't have a particularly strong opinion either way. But what I do know is that the implementation as it is now is broken as the `as` is being used for the *wrong* foreign key (again please see above).

I'm really sorry - I'm not having a go at you. As I've said, I find the associations API endlessly confusing, and so it's not surprising that some bugs have crept in.

Would be great if you have an opinion on how to fix it...

**Comment by felixfbecker** [comment]
Finally had the time to read through @overlookmotel's post, he raises valid points and I agree with point 1 and 2 of his conclusions.

**Comment by sushantdhiman** [comment]
Finally I got a chance to read the whole post. I think everything boils down to this

- Using `as` for foreign key base causing wrong / illogical / duplicate foreign key generation
- We have to revert `as` behavior for `ManyToMany` at least.

I think this course of action would make most sense to me

- Throw error when user try to use `as` without foreign key. This would solve the problems for inflected, illogical or duplicate foreign key at same time. ( Yes, I know we have to press a few more keys, but I think its worth it )
- Revert `as` usage for `ManyToMany`
- Discuss / Implement @overlookmotel suggestions from #6091 (didn't get chance to have a look yet) by looking at other frameworks

**Comment by felixfbecker** [comment]
@sushantdhiman I'm for option 2 and against 1/3.

**Comment by mickhansen** [comment]
For now, we need to revert `as` to `foreignKey` for all non-belongsTo association. We can then consider a course of action after that.

**Comment by overlookmotel** [comment]
@mickhansen OK cool. How do we do revert? Does it need a PR? Or does git have a `revert <commit sha>`-type function?

**Comment by mickhansen** [comment]
You can open revert PRs from old merged PRs

**Comment by felixfbecker** [comment]
@overlookmotel `git revert <sha of the merge/squashed commit>`

**Comment by sushantdhiman** [comment]
There is a revert button in PR, you can use it as well

**Comment by sushantdhiman** [comment]
Merged #6109
